tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Kafka/KafkaHelper.cs

```diff
@@ -40,11 +41,17 @@ internal static Scope CreateProducerScope(Tracer tracer, ITopicPartition topicPa
                     return null;
                 }
 
-                string serviceName = settings.GetServiceName(tracer, KafkaConstants.ServiceName);
+                var schema = tracer.Settings.Schema;
+                string operationName = schema.Messaging.GetOutboundOperationName(MessagingType);
+                if (!tracer.Settings.TryGetServiceName(MessagingType, out string serviceName))
+                {
+                    serviceName = schema.Messaging.GetOutboundServiceName(tracer.DefaultServiceName, MessagingType);
+                }
```

Comment:
While I generally agree with the "pure functions are better", I feel like this adds a lot of ceremony which we should try to abstract away 🤔 I know this was one of the designs you rejected, but I would probably favor something more like this:

```csharp
interface IDatabaseSchema
{
    string GetOperationName(string databaseType);
    string GetServiceName(string databaseType);
}

interface IMessagingSchema
{
    string GetInboundOperationName(string messagingSystem);
    string GetInboundServiceName(string messagingSystem);
    string GetOutboundOperationName(string messagingSystem);
    string GetOutboundServiceName(string databaseType);
}
```

and exposing these on `Tracer`. This code then becomes

```csharp
var schema = tracer.Settings.Schema;
string operationName = tracer.Schema.Messaging.GetOutboundOperationName(MessagingType);
serviceName = tracer.Schema.Messaging.GetOutboundServiceName(MessagingType);
```

or, slightly different:

```csharp
internal readonly record struct SchemaDetails(string OperationName, string ServiceName);

interface IDatabaseSchema
{
    SchemaDetails Get(string databaseType);
}

interface IMessagingSchema
{
    SchemaDetails GetInbound(string messagingSystem);
    SchemaDetails GetOutbound(string messagingSystem);
}
```

so that usage then becomes:

```csharp
var details = tracer.Settings.Schema.Messaging.GetOutbound(MessagingType);
string operationName = details.OperationName; // can inline these
string serviceName = details.ServiceName; // can inline these
```

The actual `ISchema` object would need to be stateful (as depends on tracer settings + default service name), but could live on the `TracerManager` perfectly happily. The `ISchema` object is still very testable, but it becomes a lot easier to use it IMO?

Reply:
> I feel like this adds a lot of ceremony which we should try to abstract away 🤔

I agree with this completely and this was the main internal debate I was having when establishing a preference between the approaches.

I think both of your suggestions are good, with a slight preference for the first one where we directly return strings 👍🏼

Reply:
Done in 5958c51d18164a319d6a1488076ed19f95098c04